auth/range_perm_cache.go

```diff
@@ -146,18 +87,44 @@ func checkKeyPerm(cachedPerms *unifiedRangePermissions, key, rangeEnd []byte, pe
 		plog.Panicf("unknown auth type: %v", permtyp)
 	}
 
-	requiredPerm := &rangePerm{begin: key, end: rangeEnd}
+	ivl := adt.NewStringAffineInterval(key, rangeEnd)
 
-	for _, perm := range tocheck {
-		if isSubset(requiredPerm, perm) {
-			return true
+	minBegin := adt.StringComparable("")
+	maxEnd := adt.StringComparable("\xff")
+
+	tocheck.Visit(ivl, func(n *adt.IntervalValue) bool {
+		if n.Ivl.Begin.Compare(minBegin) < 0 {
+			minBegin = n.Ivl.Begin.(adt.StringComparable)
+		}
+
+		if maxEnd.Compare(n.Ivl.End) < 0 {
+			maxEnd = n.Ivl.End.(adt.StringComparable)
 		}
+
```

Comment:
They're not necessarily continuous for cases like [a, f), [c, d), [f, z). If you're feeling adventurous, this can probably be written abstractly using only `adt.Comparable`.